### PR TITLE
feat(codedb): agent discoverability — banner framing, ergonomics, verb wrappers

### PR DESCRIPTION
**Pull request by rupakm**
## Summary

Three batches of changes that make `ox code` actually reachable by AI coworkers (Claude Code, Cursor, Codex, Windsurf, etc.). Driven by the observation that even though `ox code` exists and would help agents understand the codebase, **agents almost never reach for it** — defaulting to `Grep`/`Glob`/`Read` chains instead.

The investigation report (`docs/specs/codedb-agent-discoverability.md`) is included in this PR for the full "why." TL;DR:

1. **Framing was prescriptive, not demonstrative.** The old banner told agents to "PREFER `ox code search` over Grep" without showing what `ox code` can do that grep cannot. Pre-trained tool-use instincts (grep/rg/find) override soft `PREFER` directives.
2. **The DSL — `type:`, `calls:`, `calledby:`, `confidence:`, `before:`/`after:`, `/regex/` — is the unique value, and it was surfaced nowhere the agent could see.** Not in `--help` (zero examples, two flags), not in skill files, not in `.claude/rules/`. ADR-019's call-graph capability sat behind DSL filters with no verb-mode wrapper.
3. **Output trained the agent off the tool.** 120-char snippet truncation taught agents on first call that "ox returns less than grep." Hard errors on `code index is currently being built` taught them to abandon the tool. Stale-index anxiety was unaddressed.

### Batch 1 — Framing (3 commits, instruction-only)

- **`cmd/ox/agent_prime_xml.go`** — `<code-search>` banner rewritten from prescription to demonstration. Now lists 9 concrete DSL invocations showing what grep cannot do (`calledby:authenticate`, `calls:Handler depth:2`, `type:pr`, `type:comment ckind:todo`, etc.) and the full keyword inventory. Restricts Grep/Glob to "exact-string-in-known-file" and "0-result fallback." Lives in the **static prefix-cache region** so the +330 token cost is paid once per session.
- **`cmd/ox/code.go`** — `codeCmd.Long` and `codeSearchCmd.Long` carry full DSL grammar + 7 example queries spanning symbol / call-graph / PR / comment / history / regex intents. Previously `--help` showed two flags and zero examples.
- **`cmd/ox/code_activity.go`, `code_prs.go`** — `Short` strings reworded from pipeline-internal language ("for the fact extractor") to agent-visible value props.
- **`internal/prime/guidance.go`** — `<commands>` table expanded from 2 ox-code rows to 6, surfacing `calls`/`calledby`, `type:pr`/`type:issue`, `prs`, `activity`, `insights` as distinct intents.
- **`cmd/ox/agent_prime.go`** — `CodeSearchTip` rewording aligned with new framing.
- **`.claude/rules/ox-code.md` (new)** — decision tree + DSL cheatsheet + anti-patterns + fallback policy + index-health notes.
- **`claude-plugin/skills/ox/SKILL.md`, `extensions/claude/commands/ox.md`** — "Search Code, History, PRs" sections added. Previously silent on `ox code`.

### Batch 2 — Ergonomics (1 commit, small CLI surface changes)

- **Snippet width**: default 120 → 200 chars, plus `--snippet N` for override. 120 was cutting mid-arg-list on Go signatures. `--full-json` description updated from `~6x` to `~4x` tokens.
- **Structured index-not-ready response**: `ox code search|insights|prs|activity` now emit `{"status":"indexing"|"not_indexed", "fallback_hint":"..."}` JSON on stdout (exit 0) when invoked by an agent (`detectAgentContext()` true). Humans still get the human-readable error. Agents branch on `status` instead of abandoning the tool on stderr.
- **JIT DSL hint**: when the agent issues a bare single-term query (no DSL filters, no OR, no /regex/) and non-zero results return, append a one-line hint pointing at `type:symbol`, `calledby:`/`calls:`, `type:pr`/`type:issue`, `before:`/`after:`. Just-in-time discovery without bloating responses that already use the DSL.
- **Stderr latency line**: every `ox code search` now writes `codedb: N results in Tms (dirty overlays: K)` to stderr. Calibrates agents on latency without per-call context cost. Suppressed by `--quiet`.

### Batch 3 — Verb wrappers (1 commit + 1 follow-up, new surface)

ADR-019's resolved call graph (`calls:` / `calledby:`) was completely hidden behind DSL grammar. Agents pattern-match on verbs (`callers`, `who-calls`) more reliably than on filter keywords. New thin wrappers (each a shell over the shared `runCodeSearch` helper extracted from `codeSearchCmd.RunE`):

```
ox code defs <name>                # where is <name> defined?   (type:symbol)
ox code callers <name>             # who calls <name>?           (calledby:)
ox code callees <name> --depth N   # what <name> calls           (calls: depth:)
ox code refs <name> [--lang go]    # text references             (type:code lang:)
ox code log <path> [--author X --after YYYY-MM-DD]   # commit history (file: type:commit ...)
```

All verbs inherit the same flags as `ox code search` (`--limit`, `--snippet`, `--full-json`) so output is byte-for-byte identical with the equivalent DSL invocation. Banner, guidance table, skill, slash command, and rule all updated to list verbs first, DSL-mode second.

### Follow-ups (2 commits, review feedback)

- **`fix(codedb): sub-second precision on search-latency stderr line + compact commit fields`**
  - `formatSearchLatency` exposes µs/ms below 1s so the stderr line shows real numbers (was floored to `<1s`)
  - `compactSearchResult` adds `commit_hash`, `author`, `commit_message` — `type:commit` results used to drop these and return `{"snippet":""}`
  - `ox code log <path>` defaults `--after` to one year ago so the natural invocation always returns data (executor requires at least one of author/before/after/message)

- **`fix(codedb): review feedback — preserve snippet JSON contract, wrap banner lines`**
  - `Snippet` field reverted from `omitempty` to always-present (preserves agent JSON contract)
  - 3 banner lines wrapped to stay within ~80 chars

- **`fix(codedb): close deferred review items — rune-aware snippet truncation + verb-arg validation`**
  - `compactSnippet` now walks runes via `for byteIdx := range result` — byte-slice truncation could corrupt multi-byte UTF-8 (Greek identifiers, emoji, CJK PR titles) and break downstream JSON encoding
  - `validateSymbolArg` / `validatePathArg` reject whitespace, colons, and quote characters on verb args + `--author/--after/--before/--lang` flag values — prevents DSL-injection confusion like `ox code callers "foo calledby:bar"` rendering as `calledby:foo calledby:bar`

## Tests

26 new test cases across `cmd/ox/code_test.go` and `cmd/ox/code_verbs_test.go`:

| Suite | Coverage |
|---|---|
| `TestIsBareQuery` | 9 boundary cases (DSL detection for JIT hint) |
| `TestCompactSearchResults_*` | snippet default, override, paging guidance, JIT-hint boundary |
| `TestCompactSnippet_RuneAware_*` | ASCII / 2-byte (λ) / 4-byte (🤖) / pass-through |
| `TestEmitIndexNotReadyJSON_*` | both `indexing` and `not_indexed` shapes |
| `TestIndexStatusConstants_AreStable` | locks agent JSON contract |
| `TestValidateSymbolArg_*` | clean ids, 7 injection patterns, empty |
| `TestValidatePathArg_*` | paths, DSL-injection rejection |
| `TestVerb*_BuildsDSL` | 8 round-trip checks (verb → ParsedQuery) |
| `TestVerbCommands_AreRegistered` | all 5 verbs reach `codeCmd` |

Full suite: **5,872 pass / 2 pre-existing fails / 7 skip** across 12 packages. Two failures (`TestDoctorFreshInstall_*`) are unrelated — verified by stashing the branch and rerunning (same failures). Root cause is a daemon-cached version-update notice being treated as an unexpected warning by the test fixtures.

### Banner token cost
| | Old | New | Delta |
|---|---|---|---|
| `<code-search>` block bytes | 517 | 1,842 | +1,325 |
| Estimated tokens | ~130 | ~460 | **+330** |
| Region | prefix-cacheable | prefix-cacheable | unchanged — paid once per session, amortized across all turns under Anthropic's 5-min cache TTL |

### Search latency (sub-second precision now visible)
```
codedb: 11 results in 18ms (dirty overlays: 1)
codedb: 11 results in  3ms (dirty overlays: 1)
codedb: 11 results in  4ms (dirty overlays: 1)
```
Pure search 3–28ms vs grep ~40-100ms / ripgrep ~45-56ms. Wall-clock (incl. binary startup) is ~1s for `ox code`; that's a known property of the ox binary, not introduced by this PR.

### Subagent A/B
Two rounds of testing whether the new framing actually changes agent behavior:

**Round 1 — same task, different framing**
- `ox` repo: 5 understand-the-codebase tasks × 2 framings = 10 subagents. Aggregate: A=37 tools / 103,945 tokens / 137.7s wall, B=36 / 99,371 / 135.5s — **−3% tools, −4% tokens, −2% wall.** Small but consistent direction; B was also more accurate on T1 (no false positive caller).
- TraceForge repo (Rust, no `.sageox/` priming, 1,484 symbols): 3 tasks × 2 = 6 subagents. Aggregate: B was **worse** — +31% tools, +30% tokens, +56% wall. Direction reversed because (a) no environmental priming so framing came from prompt only, (b) symbols not in index → ox code calls returned 0 → fell back to grep anyway, (c) repo too small for the index advantage to compound.

**What this proves**: framing helps when (i) the repo is large enough that codedb actually beats grep, (ii) the target symbol exists in the index, (iii) the agent gets the banner via `ox agent prime` not just inline prompt. The initial N=1 test showing −70% over-sold the win; N=8 across two repos shows a modest, conditional improvement. Honest writeup in `docs/specs/codedb-discoverability-impl-results.md`.

## Implementation discipline

- No production logic changed in Batch 1 (instruction-only)
- `runCodeSearch` extracted from `codeSearchCmd.RunE` so verb wrappers share the index-not-ready handling, dirty-overlay attach, search, and stats emission — no duplication
- All commits follow conventional format
- ADR-019 dependencies (`calls:`/`calledby:`/`confidence:`) already in main (`12643f0`) — this PR builds on them
- Deferred items from the investigation that are NOT in this PR (documented in the report for future work):
  - `number:<n>` DSL filter (would enable `ox code pr <number>` verb)
  - `ox find` root-level alias (touches `rootCmd` — required-review surface per `AGENTS.md`)
  - Tool-use telemetry counter (explicitly excluded by user instruction during implementation)
  - Bleve symbol-index health check in `ox doctor` (separate hardening work)

## Test plan

- [x] `go build ./...` clean
- [x] `go vet ./cmd/ox/... ./internal/prime/...` clean
- [x] `go test ./cmd/ox/ ./internal/codedb/... ./internal/prime/...` — 5,872 pass / 2 pre-existing fail / 7 skip
- [x] Manual: `./ox agent prime --format=xml` shows new banner with verb-mode + DSL-mode examples
- [x] Manual: `./ox code defs <name>`, `callers`, `callees`, `refs`, `log` all execute against live index
- [x] Manual: `./ox code callers "foo calledby:bar"` rejected with helpful error
- [x] Manual: `./ox code search "λ" type:code` returns valid UTF-8 (no mid-rune corruption)
- [x] Subagent A/B (N=8 across two repos) — see `docs/specs/codedb-discoverability-impl-results.md`

## Files changed

- New: `cmd/ox/code_verbs.go`, `cmd/ox/code_verbs_test.go`, `.claude/rules/ox-code.md`, `docs/specs/codedb-agent-discoverability.md`, `docs/specs/codedb-discoverability-impl-results.md`
- Modified: `cmd/ox/code.go`, `cmd/ox/code_test.go`, `cmd/ox/code_activity.go`, `cmd/ox/code_insights.go`, `cmd/ox/code_prs.go`, `cmd/ox/agent_prime.go`, `cmd/ox/agent_prime_xml.go`, `cmd/ox/agent_query.go`, `internal/prime/guidance.go`, `claude-plugin/skills/ox/SKILL.md`, `extensions/claude/commands/ox.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: SageOx <ox@sageox.ai>

## Release Notes

* **New Features**
  * Added verb-based commands for callers, callees, definitions, references, and commit history.
  * Expanded `ox code` search across code, history, pull requests, issues, and comments.
  * Improved snippets, UTF-8 handling, paging guidance, and structured index-status responses.
  * Added clearer guidance for indexed and unavailable code searches.

* **Documentation**
  * Expanded search documentation with DSL examples, filters, fallback behavior, and index guidance.

* **Tests**
  * Added coverage for search commands, validation, formatting, and index-status responses.

### CODE REVIEW

File: cmd/ox/code.go [hunk 1]
```diff
@@ -99,74 +114,208 @@ var codeIndexCmd = &cobra.Command{
...
-			}
-		} else {
-			compact := compactSearchResults(results, limit)
-			if err := enc.Encode(compact); err != nil {
-				return fmt.Errorf("encode: %w", err)
+	if fullJSON {
+		resp := &combinedQueryResponse{CodeResults: results}
+		if err := enc.Encode(resp); err != nil {
+			return fmt.Errorf("encode: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/code.go:265]: Fixed in 3685f084. `runCodeSearch` now stats `metadata.db` before `codedb.Open` and, for agent contexts, emits `{"status":"not_indexed","fallback_hint":"..."}` (clear actionable error for humans) — mirroring the `code_activity.go` pattern. Because all five verbs (defs/callers/callees/refs/log) route through `runCodeSearch`, this closes the never-indexed gap for every one of them. Red-first tests: `TestRunCodeSearch_NeverIndexed_AgentGetsNotIndexedJSON` / `_HumanGetsClearError`.

File: cmd/ox/code.go [hunk 2]
```diff
@@ -99,74 +114,208 @@ var codeIndexCmd = &cobra.Command{
...
 
-		// attach all daemon-built dirty overlays for uncommitted file search
-		// (supports multiple simultaneous worktrees)
-		if n := db.AttachAllDirtyIndexes(); n > 0 {
-			slog.Debug("attached dirty overlays", "count", n)
-		}
+	db, err := codedb.Open(dataDir)
+	if err != nil {
+		return fmt.Errorf("open codedb: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/code.go:212]: Fixed in 3685f084. `runCodeSearch` now stats `metadata.db` before `codedb.Open` and, for agent contexts, emits `{"status":"not_indexed","fallback_hint":"..."}` (clear actionable error for humans) — mirroring the `code_activity.go` pattern. Because all five verbs (defs/callers/callees/refs/log) route through `runCodeSearch`, this closes the never-indexed gap for every one of them. Red-first tests: `TestRunCodeSearch_NeverIndexed_AgentGetsNotIndexedJSON` / `_HumanGetsClearError`.

File: cmd/ox/code_verbs.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		after, _ := cmd.Flags().GetString("after")
+		before, _ := cmd.Flags().GetString("before")
+		for _, v := range []struct{ name, val string }{
+			{"--author", author}, {"--after", after}, {"--before", before},
+		} {
+			if v.val == "" {
+				continue
+			}
+			if err := validateSymbolArg(v.name, v.val); err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/code_verbs.go]: Fixed in 3685f084. Added `validateDateArg` — it permits `:` so ISO 8601 timestamps (e.g. `2026-04-01T10:00:00Z`) pass, while still rejecting whitespace and quotes (the real DSL-injection vectors). `--after`/`--before` now route through it; `--author` keeps `validateSymbolArg`. Verified end-to-end that `ParseQuery` accepts `after:2026-04-01T10:00:00Z`. Red-first test: `TestCodeLog_AfterTimestamp_PassesValidationLoop`.

File: docs/specs/codedb-agent-discoverability.md [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+ckind: / comment-kind:<kind>
+state:<pr_state>
+confidence:{extracted,inferred,ambiguous}   # ADR-019
+OR                       # boolean across groups
+/regex/                  # forced regex
+```
+
+This is a Sourcegraph-class query DSL. **None of it is documented in
+`ox code search --help`**, which currently shows two flags (`--full-json`,
+`--limit`) and zero examples.
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply docs/specs/codedb-agent-discoverability.md:96]: Fixed in afeb6ed3. The Status line now marks this a **pre-implementation snapshot** (dated 2026-05-28) and states that the "\not documented / verbs out of scope / proposed index statuses"\ claims are the *before* picture that PR #664 implements, pointing to the impl-results doc for what shipped. One qualifier at the top covers the 88-90 / 306-325 / 350-367 sections.

File: docs/specs/codedb-discoverability-impl-results.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+warning by the test fixtures.
+
+### Live-index integration check
+
+After `ox code index --full` (526 commits, 6,818 blobs, 24,777 symbols indexed):
+
+| Command | Result | Notes |
+|---|---|---|
+| `ox code log cmd/ox/code.go --after 2026-06-01 --limit 3` | 3 results | Hash + author + message in compact JSON; this branch's own commits surface correctly |
+| `ox code log cmd/ox/code.go --limit 3` | 20 results | Defaulted `--after` to one year ago — verifies the default-window fallback |
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply docs/specs/codedb-discoverability-impl-results.md]: Fixed in afeb6ed3. Line 66 now reads "\3 shown (20 total)"\ with a note that compact JSON returns --limit entries while `total` reports all matches — consistent with the --after row above it.
